src/login/logind-session.c

```diff
@@ -182,9 +182,15 @@ int session_set_leader(Session *s, pid_t pid) {
         if (s->leader.pid == pid)
                 return 0;
 
-        r = pidref_set_pid(&pidref, pid);
-        if (r < 0)
-                return r;
+        if (pidfd >= 0) {
+                r = pidref_set_pidfd(&pidref, pidfd);
+                if (r < 0)
+                        return r;
+        } else {
+                r = pidref_set_pid(&pidref, pid);
+                if (r < 0)
+                        return r;
+        }
```

Comment:
please move the `if (r < 0) return r;` parts out of the inner blocks so that you can share it between both blocks.